design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+  2) Once a new active is chosen, depending on how far behind (lag) it is with respect to the changelog, 
+  it may take up  few seconds or even minutes before the standby has fully caught up to the last committed offset. 
+
+In total, it takes >>10 seconds before a query can start succeeding again. What can we do until a
+new active is ready to serve requests? 
+[KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance) 
+has laid the groundwork to allow us to serve requests from standbys, even if they are still in 
+rebalancing state or have not caught up to the last committed offset of the active. 
+
+Every ksql server now has the information of:
+
+1. Local offset lag: The lag of all partitions the server hosts (whether active or standby)
+2. Routing Metadata for a key: The partition, active host and standy hosts per key
+
+This information allows each server (with some communication, discussed below) to know the global 
+lag information of every topic partition whether it is the standby or active. This enables us to 
+implement a policy where `B` having established that `A` is down (after trying to send a request 
+to `A` and seeing it failed), to decide whether it (`B`) will serve it or `C`. This effectively 
+means that there will be little to no down time in serving requests at the cost of consistency as
+`B` or `C` may serve stale (not caught up) data. Hence, we achieve high availability at the cost of 
+consistency. Eventual consistency for the win!
+
+
+## What is in scope
+Ensure availability for pull queries when KSQL servers fail (with `ksql.streams.num.standby.replicas=N`, 
+we can tolerate upto N such failures) 
+
+## What is not in scope
+
+- Address unavailability caused by cold starts of KSQL server i.e when a new server is added to a 
+KSQL cluster, it must first rebuild its state off the changelog topics and that process still could 
+take a long time. 
+- Try to improve consistency guarantees provided by KSQL i.e reduce the amount of time it takes to 
+rebalance or standby replication to catch up. 
+
+## Value/Return
+
+The cluster of KSQl server will be able to serve requests even when the active is down. This mitigates 
+a large current gap in deploying KSQL pull queries for mission critical use-cases, by significantly 
+reducing failure rate of pull queries during server failures.
+
+## Public APIS
+No change to public APIs
+
+## Design
+
+The problem at hand is how to determine if a server is down and where to route a request when it is 
+down. Every KSQL server must have available the information of what other KSQL servers exist in the 
+cluster, their status and their lag per topic partition. There are three components to this: 
+Healthchecking, Lag reporting and Lag-aware routing.  
+
+### Failure detection/Healthcheck
+
+Failure detection/ healthcheck is implemented via a periodic hearbeat. KSQL servers either broadcast 
```

Comment:
Good question. Streams rebalancing itself uses the same consumer group protocol. In fact, the consumer group does membership checks using heartbeats as well, but the issue was we could not tune the heartbeat frequency down to the sub-second or few hundred ms levels we need for providing the availability SLA we need. Doing so would result in sending tons of messages to Kafka brokers and could affect other workloads/consumers.. Kafka cluster being a shared resource potentially across many ksql clusters, you can imagine it will become a scaling bottleneck.

So, we chose to move this heartbeating into ksql server, so that we can do more aggressive heartbeats, which in turn reduce failure detection time and importantly, heartbeating is just within a single ksql cluster. So it scales much better. 

Hope that clarifies.